src/firebase_functions/pubsub.py

```diff
@@ -20,69 +24,110 @@
 
 T = TypeVar("T")
 
-Request = flask.Request
-Response = flask.Response
-
 
 @dataclass(frozen=True)
 class CloudEvent(Generic[T]):
+    id: str
+    datacontenttype: str
     specversion: str
     source: str
-    subject: str
     type: str
     time: dt.datetime
     data: T
 
 
 @dataclass(frozen=True)
 class Message(Generic[T]):
+    """
+    Wrapper around a Pub/Sub message.
+    """
     message_id: str
     publish_time: dt.datetime
-    data: str
-    attributes: dict[str, str]
-    ordering_key: str
+    data: Optional[str] = None
+    attributes: Optional[dict[str, str]] = None
+    ordering_key: Optional[str] = None
 
     @property
-    def json(self) -> Dict[str, Any]:
-        return {
+    def json(self) -> Optional[T]:
+        try:
+            if self.data is not None:
+                return json.loads(base64.b64decode(self.data).decode("utf-8"))
+            else:
+                return None
+        except Exception as e:
+            raise Exception(
+                f"Unable to parse Pub/Sub message data as JSON: {e}") from e
+
+    def asdict(self) -> dict[str, Any]:
+        dict_message: dict[str, Any] = {
             "message_id": self.message_id,
-            "publish_time": self.publish_time,
             "data": self.data,
-            "attributes": self.attributes,
-            "ordering_key": self.ordering_key,
+            "publish_time": self.publish_time,
         }
 
+        if self.attributes is not None:
+            dict_message["attributes"] = self.attributes
+        if self.ordering_key is not None:
+            dict_message["ordering_key"] = self.ordering_key
 
-CloudEventMessage = CloudEvent[Message[str]]
+        return dict_message
 
 
-class MessagePublishedData(TypedDict):
-    message: object
+@dataclass(frozen=True)
+class MessagePublishedData(Generic[T]):
+    message: Message[T]
     subscription: str
 
 
 def pubsub_wrap_handler(
-    func: Callable[[CloudEventMessage], None],
-    raw: dict[str, Any],
-) -> Response:
-    event: CloudEvent[Message[str]] = CloudEvent(
-        **raw,
-        time=dt.datetime.fromisoformat(raw["time"]),
-        data=Message(
-            **raw["data"]["message"],
-            publish_time=dt.datetime.fromisoformat(
-                raw["data"]["message"]["publish_time"]
-            ),
+    func: Callable[[CloudEvent[MessagePublishedData[T]]], None],
+    raw: ce.CloudEvent,
+) -> flask.Response:
+    if isinstance(raw, ce.CloudEvent):
+        event_dict = {**raw}
+    else:
+        event_dict = raw
+
+    data = event_dict["data"]
+    message_dict = data["message"]
+
+    time = dt.datetime.strptime(
+        message_dict["publish_time"],
+        "%Y-%m-%dT%H:%M:%S.%f%z",
+    )
+
+    # Convert the UTC string into a datetime object
+    event_dict["time"] = time
+    message_dict["publish_time"] = time
```

Comment:
`time` & `publish_time` are two separate fields, lets treat them as such -  since they can be different values in some scenarios, e.g. pubsub retries or scheduled publishing

```suggestion
    # Convert the UTC strings into a datetime objects
    time = dt.datetime.strptime(
        event_dict["time"],
        "%Y-%m-%dT%H:%M:%S.%f%z",
    )
    publish_time = dt.datetime.strptime(
        message_dict["publish_time"],
        "%Y-%m-%dT%H:%M:%S.%f%z",
    )
    event_dict["time"] = time
    message_dict["publish_time"] = publish_time
```